programs/keeper-client/KeeperClient.cpp

```diff
@@ -0,0 +1,294 @@
+{
+    loadCommands({
+        {"set", 2, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             client->zookeeper->set(client->getAbsolutePath(args[1]), args[2]);
+         }},
+
+        {"create", 2, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             client->zookeeper->create(client->getAbsolutePath(args[1]), args[2], zkutil::CreateMode::Persistent);
+         }},
+
+        {"get", 1, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             std::cout << client->zookeeper->get(client->getAbsolutePath(args[1])) << "\n";
+         }},
+
+        {"ls", 0, [](KeeperClient * client, const std::vector<String> & /* args */)
+         {
+             auto children = client->zookeeper->getChildren(client->cwd);
+             for (auto & child : children)
+                 std::cout << child << " ";
+             std::cout << "\n";
+         }},
+
+        {"ls", 1, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             auto children = client->zookeeper->getChildren(client->getAbsolutePath(args[1]));
+             for (auto & child : children)
+                 std::cout << child << " ";
+             std::cout << "\n";
+         }},
+
+        {"cd", 0, [](KeeperClient * /* client */, const std::vector<String> & /* args */)
+         {
+         }},
+
+        {"cd", 1, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             auto new_path = client->getAbsolutePath(args[1]);
+             if (!client->zookeeper->exists(new_path))
+                 std::cerr << "Path " << new_path << " does not exists\n";
+             else
+                client->cwd = new_path;
+         }},
+
+        {"rm", 1, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             client->zookeeper->remove(client->getAbsolutePath(args[1]));
+         }},
+
+        {"rmr", 1, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             String path = client->getAbsolutePath(args[1]);
+             client->askConfirmation("You are going to recursively delete path " + path,
+                                     [client, path]{ client->zookeeper->removeRecursive(path); });
+         }},
+    });
+
+    String home_path;
+    const char * home_path_cstr = getenv("HOME"); // NOLINT(concurrency-mt-unsafe)
+    if (home_path_cstr)
+        home_path = home_path_cstr;
+
+    if (config().has("history-file"))
+        history_file = config().getString("history-file");
+    else
+        history_file = home_path + "/.keeper-client-history";
+
+    if (!history_file.empty() && !fs::exists(history_file))
+    {
+        try
+        {
+            FS::createFile(history_file);
+        }
+        catch (const ErrnoException & e)
+        {
+            if (e.getErrno() != EEXIST)
+                throw;
+        }
+    }
+
+    EventNotifier::init();
```

Comment:
Why do we need EventNotifier in client?

Reply:
ZooKeeper.h needs it for some reason. I didn't look up why exactly, it just said, 'EventNotifier is not initialized, it's a bug', so I turned it on.

Reply:
Here it is https://github.com/ClickHouse/ClickHouse/blob/d8cda3dbb8f9c2bfcd9bff2993490f0ed1d7499b/src/Common/ZooKeeper/ZooKeeperImpl.cpp#L885